// react-native-timer-picker/dist/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CustomTimerPickerModalStyles", {
  enumerable: true,
  get: function () {
    return _TimerPickerModal.CustomTimerPickerModalStyles;
  }
});
Object.defineProperty(exports, "CustomTimerPickerStyles", {
  enumerable: true,
  get: function () {
    return _TimerPicker2.CustomTimerPickerStyles;
  }
});
Object.defineProperty(exports, "TimerPicker", {
  enumerable: true,
  get: function () {
    return _TimerPicker.default;
  }
});
Object.defineProperty(exports, "TimerPickerModal", {
  enumerable: true,
  get: function () {
    return _components.default;
  }
});
Object.defineProperty(exports, "TimerPickerModalProps", {
  enumerable: true,
  get: function () {
    return _components.TimerPickerModalProps;
  }
});
Object.defineProperty(exports, "TimerPickerModalRef", {
  enumerable: true,
  get: function () {
    return _components.TimerPickerModalRef;
  }
});
Object.defineProperty(exports, "TimerPickerProps", {
  enumerable: true,
  get: function () {
    return _TimerPicker.TimerPickerProps;
  }
});
Object.defineProperty(exports, "TimerPickerRef", {
  enumerable: true,
  get: function () {
    return _TimerPicker.TimerPickerRef;
  }
});
var _components = _interopRequireWildcard(require("./components"));
var _TimerPicker = _interopRequireWildcard(require("./components/TimerPicker"));
var _TimerPickerModal = require("./components/TimerPickerModal.styles");
var _TimerPicker2 = require("./components/TimerPicker/TimerPicker.styles");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
