// react-native-timer-picker/dist/commonjs/components/TimerPicker/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _DurationScroll = _interopRequireDefault(require("./DurationScroll"));
var _TimerPicker = require("./TimerPicker.styles");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
const TimerPicker = /*#__PURE__*/(0, _react.forwardRef)(({
  allowFontScaling = false,
  onDurationChange,
  initialHours = 0,
  initialMinutes = 0,
  initialSeconds = 0,
  hideHours = false,
  hideMinutes = false,
  hideSeconds = false,
  hoursPickerIsDisabled = false,
  minutesPickerIsDisabled = false,
  secondsPickerIsDisabled = false,
  hourLimit,
  minuteLimit,
  secondLimit,
  hourLabel,
  minuteLabel,
  secondLabel,
  padWithNItems = 1,
  disableInfiniteScroll = false,
  aggressivelyGetLatestDuration = false,
  use12HourPicker = false,
  amLabel = "am",
  pmLabel = "pm",
  LinearGradient,
  pickerContainerProps,
  pickerGradientOverlayProps,
  topPickerGradientOverlayProps,
  bottomPickerGradientOverlayProps,
  styles: customStyles
}, ref) => {
  const checkedPadWithNItems = padWithNItems >= 0 ? Math.round(padWithNItems) : 0;
  const styles = (0, _react.useMemo)(() => (0, _TimerPicker.generateStyles)(customStyles, {
    padWithNItems: checkedPadWithNItems
  }), [checkedPadWithNItems, customStyles]);
  const [selectedHours, setSelectedHours] = (0, _react.useState)(initialHours);
  const [selectedMinutes, setSelectedMinutes] = (0, _react.useState)(initialMinutes);
  const [selectedSeconds, setSelectedSeconds] = (0, _react.useState)(initialSeconds);
  (0, _react.useEffect)(() => {
    onDurationChange === null || onDurationChange === void 0 || onDurationChange({
      hours: selectedHours,
      minutes: selectedMinutes,
      seconds: selectedSeconds
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedHours, selectedMinutes, selectedSeconds]);
  const hoursDurationScrollRef = (0, _react.useRef)(null);
  const minutesDurationScrollRef = (0, _react.useRef)(null);
  const secondsDurationScrollRef = (0, _react.useRef)(null);
  (0, _react.useImperativeHandle)(ref, () => {
    var _hoursDurationScrollR3, _minutesDurationScrol3, _secondsDurationScrol3;
    return {
      reset: options => {
        var _hoursDurationScrollR, _minutesDurationScrol, _secondsDurationScrol;
        setSelectedHours(initialHours);
        setSelectedMinutes(initialMinutes);
        setSelectedSeconds(initialSeconds);
        (_hoursDurationScrollR = hoursDurationScrollRef.current) === null || _hoursDurationScrollR === void 0 || _hoursDurationScrollR.reset(options);
        (_minutesDurationScrol = minutesDurationScrollRef.current) === null || _minutesDurationScrol === void 0 || _minutesDurationScrol.reset(options);
        (_secondsDurationScrol = secondsDurationScrollRef.current) === null || _secondsDurationScrol === void 0 || _secondsDurationScrol.reset(options);
      },
      setValue: (value, options) => {
        var _hoursDurationScrollR2, _minutesDurationScrol2, _secondsDurationScrol2;
        setSelectedHours(value.hours);
        setSelectedMinutes(value.minutes);
        setSelectedSeconds(value.seconds);
        (_hoursDurationScrollR2 = hoursDurationScrollRef.current) === null || _hoursDurationScrollR2 === void 0 || _hoursDurationScrollR2.setValue(value.hours, options);
        (_minutesDurationScrol2 = minutesDurationScrollRef.current) === null || _minutesDurationScrol2 === void 0 || _minutesDurationScrol2.setValue(value.minutes, options);
        (_secondsDurationScrol2 = secondsDurationScrollRef.current) === null || _secondsDurationScrol2 === void 0 || _secondsDurationScrol2.setValue(value.seconds, options);
      },
      latestDuration: {
        hours: (_hoursDurationScrollR3 = hoursDurationScrollRef.current) === null || _hoursDurationScrollR3 === void 0 ? void 0 : _hoursDurationScrollR3.latestDuration,
        minutes: (_minutesDurationScrol3 = minutesDurationScrollRef.current) === null || _minutesDurationScrol3 === void 0 ? void 0 : _minutesDurationScrol3.latestDuration,
        seconds: (_secondsDurationScrol3 = secondsDurationScrollRef.current) === null || _secondsDurationScrol3 === void 0 ? void 0 : _secondsDurationScrol3.latestDuration
      }
    };
  });
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, _extends({}, pickerContainerProps, {
    style: styles.pickerContainer,
    testID: "timer-picker"
  }), !hideHours ? /*#__PURE__*/_react.default.createElement(_DurationScroll.default, {
    ref: hoursDurationScrollRef,
    numberOfItems: 23,
    label: hourLabel ?? (!use12HourPicker ? "h" : undefined),
    isDisabled: hoursPickerIsDisabled,
    initialValue: initialHours,
    allowFontScaling: allowFontScaling,
    aggressivelyGetLatestDuration: aggressivelyGetLatestDuration,
    onDurationChange: setSelectedHours,
    pickerGradientOverlayProps: pickerGradientOverlayProps,
    topPickerGradientOverlayProps: topPickerGradientOverlayProps,
    bottomPickerGradientOverlayProps: bottomPickerGradientOverlayProps,
    disableInfiniteScroll: disableInfiniteScroll,
    padWithNItems: checkedPadWithNItems,
    limit: hourLimit,
    LinearGradient: LinearGradient,
    is12HourPicker: use12HourPicker,
    amLabel: amLabel,
    pmLabel: pmLabel,
    styles: styles,
    testID: "duration-scroll-hour"
  }) : null, !hideMinutes ? /*#__PURE__*/_react.default.createElement(_DurationScroll.default, {
    ref: minutesDurationScrollRef,
    numberOfItems: 59,
    label: minuteLabel ?? "m",
    isDisabled: minutesPickerIsDisabled,
    initialValue: initialMinutes,
    allowFontScaling: allowFontScaling,
    aggressivelyGetLatestDuration: aggressivelyGetLatestDuration,
    onDurationChange: setSelectedMinutes,
    padNumbersWithZero: true,
    pickerGradientOverlayProps: pickerGradientOverlayProps,
    topPickerGradientOverlayProps: topPickerGradientOverlayProps,
    bottomPickerGradientOverlayProps: bottomPickerGradientOverlayProps,
    disableInfiniteScroll: disableInfiniteScroll,
    padWithNItems: checkedPadWithNItems,
    limit: minuteLimit,
    LinearGradient: LinearGradient,
    styles: styles,
    testID: "duration-scroll-minute"
  }) : null, !hideSeconds ? /*#__PURE__*/_react.default.createElement(_DurationScroll.default, {
    ref: secondsDurationScrollRef,
    numberOfItems: 59,
    label: secondLabel ?? "s",
    isDisabled: secondsPickerIsDisabled,
    initialValue: initialSeconds,
    allowFontScaling: allowFontScaling,
    aggressivelyGetLatestDuration: aggressivelyGetLatestDuration,
    onDurationChange: setSelectedSeconds,
    padNumbersWithZero: true,
    pickerGradientOverlayProps: pickerGradientOverlayProps,
    topPickerGradientOverlayProps: topPickerGradientOverlayProps,
    bottomPickerGradientOverlayProps: bottomPickerGradientOverlayProps,
    disableInfiniteScroll: disableInfiniteScroll,
    padWithNItems: checkedPadWithNItems,
    limit: secondLimit,
    LinearGradient: LinearGradient,
    styles: styles,
    testID: "duration-scroll-second"
  }) : null);
});
var _default = exports.default = /*#__PURE__*/_react.default.memo(TimerPicker);
