// react-native-timer-picker/dist/commonjs/components/TimerPicker/TimerPicker.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateStyles = void 0;
var _reactNative = require("react-native");
/* eslint-disable @typescript-eslint/no-explicit-any */

const DARK_MODE_BACKGROUND_COLOR = "#232323";
const DARK_MODE_TEXT_COLOR = "#E9E9E9";
const LIGHT_MODE_BACKGROUND_COLOR = "#F1F1F1";
const LIGHT_MODE_TEXT_COLOR = "#1B1B1B";
const generateStyles = (customStyles, options) => {
  var _customStyles$pickerL, _customStyles$text, _customStyles$pickerI, _customStyles$text2, _customStyles$pickerI2, _customStyles$pickerI3, _customStyles$pickerI4;
  return _reactNative.StyleSheet.create({
    pickerContainer: {
      flexDirection: "row",
      marginRight: "8%",
      backgroundColor: (customStyles === null || customStyles === void 0 ? void 0 : customStyles.backgroundColor) ?? ((customStyles === null || customStyles === void 0 ? void 0 : customStyles.theme) === "dark" ? DARK_MODE_BACKGROUND_COLOR : LIGHT_MODE_BACKGROUND_COLOR),
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.pickerContainer)
    },
    pickerLabelContainer: {
      position: "absolute",
      right: 4,
      top: 0,
      bottom: 0,
      justifyContent: "center",
      minWidth: ((customStyles === null || customStyles === void 0 || (_customStyles$pickerL = customStyles.pickerLabel) === null || _customStyles$pickerL === void 0 ? void 0 : _customStyles$pickerL.fontSize) ?? (customStyles === null || customStyles === void 0 || (_customStyles$text = customStyles.text) === null || _customStyles$text === void 0 ? void 0 : _customStyles$text.fontSize) ?? 25) * 0.65,
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.pickerLabelContainer)
    },
    pickerLabel: {
      fontSize: 18,
      fontWeight: "bold",
      marginTop: ((customStyles === null || customStyles === void 0 || (_customStyles$pickerI = customStyles.pickerItem) === null || _customStyles$pickerI === void 0 ? void 0 : _customStyles$pickerI.fontSize) ?? (customStyles === null || customStyles === void 0 || (_customStyles$text2 = customStyles.text) === null || _customStyles$text2 === void 0 ? void 0 : _customStyles$text2.fontSize) ?? 25) / 6,
      color: (customStyles === null || customStyles === void 0 ? void 0 : customStyles.theme) === "dark" ? DARK_MODE_TEXT_COLOR : LIGHT_MODE_TEXT_COLOR,
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.text),
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.pickerLabel)
    },
    pickerItemContainer: {
      flexDirection: "row",
      height: 50,
      justifyContent: "center",
      alignItems: "center",
      width: ((customStyles === null || customStyles === void 0 || (_customStyles$pickerI2 = customStyles.pickerItem) === null || _customStyles$pickerI2 === void 0 ? void 0 : _customStyles$pickerI2.fontSize) ?? 25) * 3.6,
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.pickerItemContainer)
    },
    pickerItem: {
      textAlignVertical: "center",
      fontSize: 25,
      color: (customStyles === null || customStyles === void 0 ? void 0 : customStyles.theme) === "dark" ? DARK_MODE_TEXT_COLOR : LIGHT_MODE_TEXT_COLOR,
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.text),
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.pickerItem)
    },
    pickerAmPmContainer: {
      position: "absolute",
      right: 0,
      top: 0,
      bottom: 0,
      justifyContent: "center",
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.pickerLabelContainer),
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.pickerAmPmContainer)
    },
    pickerAmPmLabel: {
      fontSize: 18,
      fontWeight: "bold",
      marginTop: ((customStyles === null || customStyles === void 0 || (_customStyles$pickerI3 = customStyles.pickerItem) === null || _customStyles$pickerI3 === void 0 ? void 0 : _customStyles$pickerI3.fontSize) ?? 25) / 6,
      color: (customStyles === null || customStyles === void 0 ? void 0 : customStyles.theme) === "dark" ? DARK_MODE_TEXT_COLOR : LIGHT_MODE_TEXT_COLOR,
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.text),
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.pickerLabel),
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.pickerAmPmLabel)
    },
    disabledPickerContainer: {
      opacity: 0.4,
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.disabledPickerContainer)
    },
    disabledPickerItem: {
      opacity: 0.2,
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.disabledPickerItem)
    },
    pickerGradientOverlay: {
      position: "absolute",
      left: 0,
      right: 0,
      height: options.padWithNItems === 0 ? "30%" : ((customStyles === null || customStyles === void 0 || (_customStyles$pickerI4 = customStyles.pickerItemContainer) === null || _customStyles$pickerI4 === void 0 ? void 0 : _customStyles$pickerI4.height) ?? 50) * 0.8,
      ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.pickerGradientOverlay)
    }
  });
};
exports.generateStyles = generateStyles;
