// react-native-timer-picker/dist/module/components/TimerPickerModal.styles.js
/* eslint-disable @typescript-eslint/no-explicit-any */
import { StyleSheet } from "react-native";
const DARK_MODE_BACKGROUND_COLOR = "#232323";
const DARK_MODE_TEXT_COLOR = "#E9E9E9";
const LIGHT_MODE_BACKGROUND_COLOR = "#F1F1F1";
const LIGHT_MODE_TEXT_COLOR = "#1B1B1B";
export const generateStyles = customStyles => StyleSheet.create({
  container: {
    justifyContent: "center",
    alignItems: "center",
    overflow: "hidden",
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.container)
  },
  contentContainer: {
    backgroundColor: (customStyles === null || customStyles === void 0 ? void 0 : customStyles.backgroundColor) ?? (customStyles === null || customStyles === void 0 ? void 0 : customStyles.theme) === "dark" ? DARK_MODE_BACKGROUND_COLOR : LIGHT_MODE_BACKGROUND_COLOR,
    justifyContent: "center",
    alignItems: "center",
    borderRadius: 20,
    padding: 20,
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.contentContainer)
  },
  buttonContainer: {
    flexDirection: "row",
    marginTop: 25,
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.buttonContainer)
  },
  button: {
    marginHorizontal: 12,
    paddingVertical: 10,
    paddingHorizontal: 20,
    borderWidth: 1,
    borderRadius: 10,
    fontSize: 16,
    overflow: "hidden",
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.text),
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.button)
  },
  cancelButton: {
    borderColor: "gray",
    color: (customStyles === null || customStyles === void 0 ? void 0 : customStyles.theme) === "dark" ? DARK_MODE_TEXT_COLOR : "gray",
    backgroundColor: (customStyles === null || customStyles === void 0 ? void 0 : customStyles.theme) === "dark" ? "gray" : undefined,
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.text),
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.cancelButton)
  },
  confirmButton: {
    borderColor: "green",
    color: (customStyles === null || customStyles === void 0 ? void 0 : customStyles.theme) === "dark" ? DARK_MODE_TEXT_COLOR : "green",
    backgroundColor: (customStyles === null || customStyles === void 0 ? void 0 : customStyles.theme) === "dark" ? "green" : undefined,
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.text),
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.confirmButton)
  },
  modalTitle: {
    fontSize: 24,
    fontWeight: "bold",
    marginBottom: 15,
    color: (customStyles === null || customStyles === void 0 ? void 0 : customStyles.theme) === "dark" ? DARK_MODE_TEXT_COLOR : LIGHT_MODE_TEXT_COLOR,
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.text),
    ...(customStyles === null || customStyles === void 0 ? void 0 : customStyles.modalTitle)
  }
});
